internal/worker/bootstrap/addressfinder.go

```diff
@@ -5,110 +5,68 @@ package bootstrap
-	return network.NewMachineAddresses([]string{"localhost"}).AsProviderAddresses(), nil
-}
-
-// CAASNewEnviron returns a new environ for CAAS. We know that currently CAAS
-// doesn't implement InstanceListener, so we return an error.
-func CAASNewEnviron(ctx context.Context, params environs.OpenParams) (environs.Environ, error) {
-	return nil, errors.NotSupportedf("new environ")
-}
-
-// CAASBootstrapAddresses returns the bootstrap addresses for CAAS. We know that
-// currently CAAS doesn't implement InstanceListener, so we return an error.
-func CAASBootstrapAddresses(ctx context.Context, env environs.Environ, bootstrapInstanceID instance.Id) (network.ProviderAddresses, error) {
-	return nil, errors.NotSupportedf("bootstrap addresses")
-}
-
-// IAASBootstrapAddressFinder returns the bootstrap addresses for IAAS.
-func IAASBootstrapAddressFinder(ctx context.Context, config BootstrapAddressesConfig) (network.ProviderAddresses, error) {
-	env, err := getEnviron(ctx, config.SystemState, config.CloudService, config.CredentialService, config.NewEnvironFunc)
+// BootstrapAddressFinderFunc is responsible for finding the network provider
+// addresses for a bootstrap instance.
+type BootstrapAddressFinderFunc func(context.Context, instance.Id) (network.ProviderAddresses, error)
+
+// getInstanceAddresses retrieves the instance from the instance lister and
+// returns the addresses associated with the instance.
+func getInstanceAddresses(
+	ctx context.Context,
+	lister environs.InstanceLister,
+	instanceID instance.Id,
+) (network.ProviderAddresses, error) {
+	callCtx := envcontext.WithoutCredentialInvalidator(ctx)
+	instances, err := lister.Instances(callCtx, []instance.Id{instanceID})
 	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	addresses, err := config.BootstrapAddressesFunc(ctx, env, config.BootstrapInstanceID)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	return addresses, nil
-}
-
-// getEnviron creates a new environ using the provided NewEnvironFunc from the
-// worker config.
-func getEnviron(ctx context.Context, state SystemState, cloudService CloudService, credentialService CredentialService, newEnviron NewEnvironFunc) (environs.Environ, error) {
-	controllerModel, err := state.Model()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-
-	cred, err := getEnvironCredential(ctx, controllerModel, credentialService)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	cloud, err := cloudService.Cloud(ctx, controllerModel.CloudName())
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-
-	cloudSpec, err := cloudspec.MakeCloudSpec(*cloud, controllerModel.CloudRegion(), cred)
-	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, fmt.Errorf(
+			"cannot get instance %q from instance lister: %w",
+			instanceID, err,
+		)
 	}
-
-	controllerModelConfig, err := controllerModel.Config()
+	addrs, err := instances[0].Addresses(callCtx)
```

Comment:
Add a guard around this raw access so we don't panic by accident.

Reply:
Done & Done